### PR TITLE
Fix crash in DL building

### PR DESCRIPTION
**Pull request by pyfisch**
Fix one crash and some style changes.

This HTML crashed servo before. Key parts are `background-clip: content-box` and `direction: rtl`

```html
<!DOCTYPE html>
<html>
    <style>
        #span1 {
            background-clip: content-box;
        }
        #span2
        {
            direction: rtl;
        }
    </style>
    <span id="span1">Filler Text <span id="span2">txeT relliF</span></span>
</html>
```
Should I add this as a test? And where do I put this "does-it-crash?" test?

I find always passing rectangles by value a lot easier as it avoids many references and dereferences and I assume that the compiler will always use the faster one either way. If you don't like the change feel free to only merge the first commit.

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/20752)

### CODE REVIEW

File: components/layout/display_list/builder.rs [hunk 1]
```diff
@@ -840,33 +840,19 @@ impl FragmentDisplayListBuilding for Fragment {
...
-                let border_padding = self.border_padding.to_physical(style.writing_mode);
-                bounds = bounds.inner_rect(border_padding);
-                border_radii = calculate_inner_border_radii(border_radii, border_padding);
-            },
-        }
+        let color_clip = *get_cyclic(&background.background_clip.0, last_background_image_index);
+        let (bounds, border_radii) = compute_background_clip(
+            color_clip,
+            *absolute_bounds,
+            style.logical_border_width().to_physical(style.writing_mode),
```
Comment (emilio) [inline components/layout/display_list/builder.rs]: As a followup, let's double-check that computing the background-clip in terms of the padding of a fragment, but the border of another one is the right thing to do. Looks fishy.

**Review by emilio — APPROVED** [review]
Code change itself looks good, but yeah, let's add a WPT test for this so it doesn't regress. `tests/wpt/mozilla` should work for these situations.

**Comment by pyfisch** [comment]
Added the test.

File: components/layout/display_list/builder.rs (hunk 1 above)
Comment (pyfisch) [reply components/layout/display_list/builder.rs]: > Looks fishy.

It probably is. For block fragments it does not make a difference as it is always the same fragment but for inlines it is probably wrong. However I don't know how to correctly compute border_padding for inlines. I assumed it was padding + border for all four sides but [there is more going on](https://github.com/pyfisch/servo/blob/0ddcf33c97c1a1ca2ec26f5e453fc327b35a2f59/components/layout/fragment.rs#L1272-L1313).

File: tests/wpt/mozilla/tests/css/background_border_padding_crash.html [hunk 2]
```diff
@@ -0,0 +1,14 @@
+<!DOCTYPE html>
+<html>
+    <title>Assure that different writing modes do not crash background building in Servo</title>
```
Comment (emilio) [inline tests/wpt/mozilla/tests/css/background_border_padding_crash.html]: This isn't run as a test. This needs a testharness `<script>` and something like:

```
async_test(function(t) {
    window.onload = t.step_func_done();
}, "Didn't crash");
```

Comment (jdm) [reply tests/wpt/mozilla/tests/css/background_border_padding_crash.html]: Can we write it as a reference test instead?

Comment (pyfisch) [reply tests/wpt/mozilla/tests/css/background_border_padding_crash.html]: @jdm How do I best convert this to a reference test as I don't have a reference file?

Comment (jdm) [reply tests/wpt/mozilla/tests/css/background_border_padding_crash.html]: Wouldn't the reference file look like ` <span>Filler Text Filler Text</span>`?

**Comment by pyfisch** [comment]
Ok. I've added a reference to the test. Can you have a look?

**Comment by jdm** [comment]
@bors-servo retry
* #20734

**Pull request merged**